### PR TITLE
Support for Regularization

### PR DESCRIPTION
**Pull request by Ayush-iitkgp**
This implements support for [regularization](https://github.com/JuliaML/PenaltyFunctions.jl/blob/master/README.md#available-penalties) terms upported in [PenaltyFunctions.jl](https://github.com/JuliaML/PenaltyFunctions.jl/blob/master/README.md).

### CODE REVIEW

File: src/build_loss_objective.jl [hunk 1]
```diff
@@ -23,12 +23,17 @@ function build_loss_objective(prob::DEProblem,alg,loss;mpg_autodiff = false,
       end
     end
     tmp_prob = prob_generator(prob,p)
-    if typeof(loss) <: Union{CostVData,L2Loss}
+    if typeof(loss) <: Union{CostVData,L2Loss,MaximumLikelihood}
```
Comment (ChrisRackauckas) [inline src/build_loss_objective.jl]: MaximumLikelihood isn't in this PR.

File: src/cost_functions.jl [hunk 2]
```diff
@@ -1,3 +1,15 @@
+export DECostFunction, CostVData, L2Loss, MaximumLikelihood, Regularization
```
Comment (ChrisRackauckas) [inline src/cost_functions.jl]: MaximumLikelihood isn't in this PR.

**Comment by ChrisRackauckas** [comment]
Are there tests you're adding to this?

**Comment by Ayush-iitkgp** [comment]
I wonder what the test would be like?
I can add the documentation on how to use it.

**Comment by Ayush-iitkgp** [comment]
Tests failing? Looks like travis setup has some problems.

**Comment by ChrisRackauckas** [comment]
It needs PenaltyFunctions in the REQUIRE file.

**Comment by Ayush-iitkgp** [comment]
Ready to merge?

**Comment by ChrisRackauckas** [comment]
No tests on this? Should at least show that some common methods still converge to the correct parameters with this. Have you created any plots of the 1D cost surface to show that it indeed smooths it? I don't think that is possible to do as a unit test but it would be a good check to have around.

File: src/cost_functions.jl [hunk 3]
```diff
@@ -1,3 +1,15 @@
+export DECostFunction, CostVData, L2Loss, Regularization
+
+type Regularization{L,P} <: DECostFunction
+  λ::L
+  penalty::P
+end
+
+function (f::Regularization)(p)
```
Comment (Ayush-iitkgp) [inline src/cost_functions.jl]: MahalanobisPenalty works on arrays so if the parameter estimation problem is to find a single parameter, it spits out the error. Any workaround?

Comment (ChrisRackauckas) [inline src/cost_functions.jl]: I don't think we need to work around it. If that penalty is used on a scalar it's just used incorrectly.

**Comment by ChrisRackauckas** [comment]
Nice. I resolved the conflict added due to a printing change. Will merge when tests pass.

**Pull request merged**